### PR TITLE
[Feature] Alias & group call added.

### PR DESCRIPTION
**Pull request by pallabmaiti**
## Description of the change

> Alias & group call added.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java [hunk 1]
```diff
@@ -147,6 +147,32 @@ public void identify(String userId, ReadableMap traits, ReadableMap options) {
...
     }
 
+    @ReactMethod
+    public void alias(String newId, ReadableMap options) {
+        if (rudderClient == null) {
+            RudderLogger.logWarn("Dropping the Alias call as RudderClient is not initialized yet, Please use `await` keyword with the setup call");
+            return;
+        }
+        if (TextUtils.isEmpty(newId)) {
+            rudderClient.alias(Utility.convertReadableMapToOptions(options));
```
Comment (desusai7) [inline android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java:157]: We are not having an alias(options) api in Android, can you please change it to use a valid alias API ?

File: android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java [hunk 2]
```diff
@@ -147,6 +147,32 @@ public void identify(String userId, ReadableMap traits, ReadableMap options) {
...
+    }
+
+    @ReactMethod
+    public void group(String groupId, ReadableMap traits, ReadableMap options) {
+        if (rudderClient == null) {
+            RudderLogger.logWarn("Dropping the Group call as RudderClient is not initialized yet, Please use `await` keyword with the setup call");
+            return;
+        }
+        if (TextUtils.isEmpty(groupId)) {
+            rudderClient.group(Utility.convertReadableMapToTraits(traits), Utility.convertReadableMapToOptions(options));
```
Comment (desusai7) [inline android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java:170]: We are not having group(traits, options) API as well in Android, can you please change it to use a valid Group API?

File: ios/RNRudderSdkModule.m [hunk 3]
```diff
@@ -114,6 +114,36 @@ - (dispatch_queue_t)methodQueue
...
+RCT_EXPORT_METHOD(alias:(NSString*)_newId options:(NSDictionary*)_options)
+{
+    if ([RSClient sharedInstance] == nil)
+    {
+        [RSLogger logWarn:@"Dropping the Alias call as RudderClient is not initialized yet, Please use `await` keyword with the setup call"];
+        return;
+    }
+    if([_newId isEqual:@""])
+    {
+        [[RSClient sharedInstance] alias:nil options:[self getRudderOptionsObject:_options]];
```
Comment (desusai7) [inline ios/RNRudderSdkModule.m:126]: We shouldn't support making alias with `newId` as nil/empty string, as the important property we need in an alias call is the `newId` which is supposed to replace the already `existingId`.

May be we can check if the `newId` is an empty string on the React Native module only and return if its invalid.

We should check the same for Android as well.

File: ios/RNRudderSdkModule.m [hunk 4]
```diff
@@ -114,6 +114,36 @@ - (dispatch_queue_t)methodQueue
...
+RCT_EXPORT_METHOD(group:(NSString*)_groupId traits:(NSDictionary*)_traits options:(NSDictionary*)_options)
+{
+    if ([RSClient sharedInstance] == nil)
+    {
+        [RSLogger logWarn:@"Dropping the Group call as RudderClient is not initialized yet, Please use `await` keyword with the setup call"];
+        return;
+    }
+    if([_groupId isEqual:@""])
+    {
+        [[RSClient sharedInstance] group:nil traits:_traits options:[self getRudderOptionsObject:_options]];
```
Comment (desusai7) [inline ios/RNRudderSdkModule.m:141]: We shouldn't support making group with `groupId` as nil/empty string, as it wouldn't be of much use of a group call with no groupId but group traits, as they are not persisted and lost in the subsequent events.

Maybe we can check if the `groupId` is an empty string on the React Native module itself and return if it's invalid.

We should check the same for Android as well.

File: src/RudderClient.ts [hunk 5]
```diff
@@ -193,19 +193,26 @@ async function group(groupId: string, traits: Object | null = null) {
...
     return;
   }
-  logWarn("group: Method not supported");
+  bridge.group(groupId, traits, options)
 }
 
-async function alias(previousId: string, userId: string) {
-  if (previousId == undefined) {
-    logWarn("alias: Mandatory field 'previousId' missing");
+async function alias(previousId: string, userId: string): Promise<void>;
```
Comment (desusai7) [inline src/RudderClient.ts:199]: `alias(previousId, userId)` is not adhering to our spec and also we are not using the value of `previousId` anywhere in the implementation. Can we remove it ?

File: src/RudderClient.ts [hunk 6]
```diff
@@ -193,19 +193,26 @@ async function group(groupId: string, traits: Object | null = null) {
...
     return;
   }
-  logWarn("alias: Method not supported");
+  if (typeof options == "string") {
+    bridge.alias(options, null)
+  } else if (typeof options == "object") {
+    bridge.alias(newId, options)
+  } else {
+    bridge.alias(newId, null)
+  }
```
Comment (desusai7) [inline src/RudderClient.ts:215]: Can we combine the condition check for string and else into a single block and change it to the following way?

```
if (typeof options == "object") {
    bridge.alias(newId, options)
  } else {
    bridge.alias(newId, null)
  }
```

**Review by desusai7 — CHANGES_REQUESTED** [review]

File: src/RudderClient.ts [hunk 7]
```diff
@@ -193,19 +193,26 @@ async function group(groupId: string, traits: Object | null = null) {
...
     return;
   }
-  if (typeof previousId != "string") {
-    logWarn("alias: 'previousID' must be a string");
+  if (typeof newId != "string") {
+    logWarn("alias: 'newId' must be a string");
     return;
   }
-  logWarn("alias: Method not supported");
+  if (typeof options == "string") {
```
Comment (desusai7) [inline src/RudderClient.ts:209]: I don't think we need to check if `options` passed is a `string` type

In this case if both `newId` and and `options` are sent as a string, as following:
```
alias("newId1", "optionsString");
```

then the final resultant call would be 
```
bridge.alias("optionsString");
```
which is not an expected behavior

File: android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java (hunk 1 above)
Comment (pallabmaiti) [reply android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java:157]: We do have [alias(option)](https://github.com/rudderlabs/rudder-sdk-android/blob/10ff3da5112979f1c1f82115ecf35d6665e9a319/core/src/main/java/com/rudderstack/android/sdk/core/RudderClient.java#L500) API. Right?

Comment (desusai7) [reply android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java:157]: Yeah, we had alias(newId, option) API, but not alias(option).

Comment (pallabmaiti) [reply android/src/main/java/com/rudderstack/react/android/RNRudderSdkModule.java:157]: Oh, okay.

File: src/RudderClient.ts (hunk 5 above)
Comment (pallabmaiti) [reply src/RudderClient.ts:199]: The API was present earlier. We are not sure if anyone is using it or not.! Updating to this version can break their implementation. We should be compatible with the previous version.

File: src/RudderClient.ts (hunk 7 above)
Comment (pallabmaiti) [reply src/RudderClient.ts:209]: We have to for backward compatibility.

File: src/RudderClient.ts (hunk 5 above)
Comment (desusai7) [reply src/RudderClient.ts:201]: Got it.

File: src/RudderClient.ts (hunk 7 above)
Comment (desusai7) [reply src/RudderClient.ts:211]: Got it

**Review by desusai7 — APPROVED** [review]

**Pull request merged**